tools/dkcm/clerk/clerk.go

```diff
@@ -75,77 +76,199 @@ type ClerkOperations interface {
 	ListRequests(window time.Duration) ([]Request, error)
 }
 
+// compatible with both Row and Rows
+type scannable interface {
+	Scan(dest ...interface{}) error
+}
+
 // DB holds an active database connection. This implements all the fuunctions of ClerkOperations.
 type DB struct {
 	*sql.DB
 }
 
 func (c Cluster) String() string {
-	return ""
+	return fmt.Sprintf("Cluster Info: (ProjectID: %s, NodesCount: %v, NodeType: %s, Status: %s, Zone: %s)",
+		c.ProjectID, c.Nodes, c.NodeType, c.Status, c.Zone)
+}
+
+func (r Request) String() string {
+	return fmt.Sprintf("Request Info: (RequestTime: %v, NodesCount: %v, NodeType: %s, ProwJobID: %s, ClusterID: %d,Zone: %s)",
+		r.requestTime, r.Nodes, r.NodeType, r.ProwJobID, r.ClusterID, r.Zone)
+}
+
+// Populate fields of Cluster
+func populateCluster(sc scannable) (*Cluster, error) {
+	c := &Cluster{}
+	err := sc.Scan(&c.ClusterID, &c.ProjectID, &c.Nodes, &c.NodeType, &c.Status, &c.Zone)
+	return c, err
+}
+
+// Populate fields of Request
+func populateRequest(sc scannable) (*Request, error) {
+	r := &Request{}
+	err := sc.Scan(&r.accessToken, &r.requestTime, &r.Nodes, &r.NodeType, &r.ProwJobID, &r.ClusterID, &r.Zone)
+	return r, err
 }
 
 // NewDB returns the DB object with an active database connection
 func NewDB(c *mysql.DBConfig) (*DB, error) {
-	return nil, nil
+	db, err := c.Connect()
+	return &DB{db}, err
 }
 
-func (db *DB) CheckAvail() (bool, string, string) {
-	return true, "", ""
+// check available clusters in Cluster db and return zone and projectID that would be used by ProwJob
+func (db *DB) CheckAvail(nodes int, nodeType string, zone string) (bool, int64) {
+	// check whether available cluster exists
+	row := db.QueryRow("SELECT * FROM Cluster WHERE Status = 'Ready' AND Nodes = ? AND NodeType = ? AND Zone = ?",
+		nodes, nodeType, zone)
+	c, err := populateCluster(row)
+	// no available cluster is found
+	if err != nil {
+		return false, 0
+	}
+	return true, c.ClusterID
+
 }
 
 // insert a cluster entry into db
-func (db *DB) InsertCluster(nodes int, nodeType string, accessToken string, projectID string, prowJobID string, status string, zone string) error {
-	return nil
+func (db *DB) InsertCluster(nodes int, nodeType string, projectID string, status string, zone string) error {
+	stmt, err := db.Prepare(`INSERT INTO Cluster(Nodes, NodeType, ProjectID, Status, Zone)
+							VALUES (?,?,?,?,?,?)`)
+	defer stmt.Close()
+
+	if err != nil {
+		return err
+	}
+
+	_, err = stmt.Exec(nodes, nodeType, projectID, status, zone)
+	return err
+}
+
+func updateQueryString(dbName string, id int, fields []string, values []interface{}) string {
+	updateStatement := fmt.Sprintf(`UPDATE %s SET `, dbName)
+	idStatement := fmt.Sprintf(` WHERE ID = % d`, id)
+	var fieldStatements []string
+	for idx, val := range fields {
+		fieldStatements = append(fieldStatements, fmt.Sprintf("%s = %v", val, values[idx]))
+	}
+	return updateStatement + strings.Join(fieldStatements, `, `) + idStatement
```

Comment:
Using `fmt.Sprintf` can avoid some of the leading and trailing space in previous strings, which could be easier to maintain in the future